util/include/util/odbtransaction.h

```diff
@@ -175,12 +175,22 @@ void persistAll(Cont& cont_, std::shared_ptr<odb::database> db_)
         << item->toString();
       LOG(warning)
         << ex.what() << std::endl
-        << "AST nodes in this translation unit will be ignored!";
+        << "Further changes in this transaction will might be ignored!";
     }
     catch (const odb::database_exception& ex)
     {
-      // TODO: Error code should be checked and rethrow if it is not unique
-      // constraint error. Error code may be database specific.
+      LOG(debug) << item->toString();
+
+#ifdef DATABASE_PGSQL
+      if (std::string(ex.what()).find("25P02") != std::string::npos)
+      {
+        LOG(error) << "The transaction was aborted due to previous error, omitting further changes!";
```

Comment:
What does "previous error" mean?

Reply:
A previous exception. In case of PostgreSQL if a database exception is raised during a transaction, the transaction **must be rolled back**, otherwise all further commands will raise a specific exception with code **25P02**, stating that the transaction has been aborted.

Reply:
I mean does the user get notified about that exception? This is an error level log, and the exception message is not printed earlier.

Reply:
Yes.